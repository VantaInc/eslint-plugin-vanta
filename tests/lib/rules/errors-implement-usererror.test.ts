import { GraphQLRuleTester } from "@graphql-eslint/eslint-plugin";
import rule from "../../../lib/rules/errors-implement-usererror";

const ruleTester = new GraphQLRuleTester();

ruleTester.runGraphQLTests("errors-extend-usererror", rule, {
  valid: [
    {
      code: `
type BaseUserError implements UserError {
  message: String!
  code: ErrorCode!
}
`,
    },
    {
      code: `
type PasswordTooWeakError implements UserError {
  message: String!
  code: ErrorCode!
  passwordRules: [String!]!
}
`,
    },
    {
      code: `
type NonErrorType implements SomethingElse {
  cat: String!
}
`,
    },
    {
      code: `
type PasswordTooWeakError implements UserError & OtherType {
  message: String!
  code: ErrorCode!
  passwordRules: [String!]!
}
`,
    },
    {
      code: `
type NotAnErrorType {
  otherStuff: String!
}
`,
    },
    {
      code: `
extend type NotErrorType implements SomethingElse {
  passwordRules: [String!]!
}
`,
    },
    {
      code: `
extend type Error implements SomethingElse {
  passwordRules: [String!]!
}
`,
    },
  ],
  invalid: [
    {
      code: `
type NotAnErrorType implements UserError {
  message: String!
  code: ErrorCode!
  otherStuff: String!
}
`,
      errors: 1,
    },
    {
      code: `
type NotAnErrorType implements OtherType & UserError {
  message: String!
  code: ErrorCode!
  otherStuff: String!
}
`,
      errors: 1,
    },

    {
      code: `
type PasswordTooWeakError {
  passwordRules: [String!]!
}
`,
      errors: 1,
    },
    {
      code: `
type PasswordTooWeakError implements OtherType {
  passwordRules: [String!]!
}
`,
      errors: 1,
    },
    {
      code: `
extend type NotErrorType implements UserError {
  passwordRules: [String!]!
}
`,
      errors: 1,
    },
  ],
});
