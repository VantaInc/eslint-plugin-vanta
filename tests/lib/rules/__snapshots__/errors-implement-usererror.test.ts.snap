// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Invalid #1 1`] = `
"#### ⌨️ Code

      1 | type NotAnErrorType implements UserError {
      2 |   message: String!
      3 |   code: ErrorCode!
      4 |   otherStuff: String!
      5 | }

#### ❌ Error

    > 1 | type NotAnErrorType implements UserError {
        | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    > 2 |   message: String!
        | ^^^^^^^^^^^^^^^^^^
    > 3 |   code: ErrorCode!
        | ^^^^^^^^^^^^^^^^^^
    > 4 |   otherStuff: String!
        | ^^^^^^^^^^^^^^^^^^
    > 5 | }
        | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Types that do not end in 'Error' cannot implement UserError"
`;

exports[`Invalid #2 1`] = `
"#### ⌨️ Code

      1 | type NotAnErrorType implements OtherType & UserError {
      2 |   message: String!
      3 |   code: ErrorCode!
      4 |   otherStuff: String!
      5 | }

#### ❌ Error

    > 1 | type NotAnErrorType implements OtherType & UserError {
        | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    > 2 |   message: String!
        | ^^^^^^^^^^^^^^^^^^
    > 3 |   code: ErrorCode!
        | ^^^^^^^^^^^^^^^^^^
    > 4 |   otherStuff: String!
        | ^^^^^^^^^^^^^^^^^^
    > 5 | }
        | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Types that do not end in 'Error' cannot implement UserError"
`;

exports[`Invalid #3 1`] = `
"#### ⌨️ Code

      1 | type PasswordTooWeakError {
      2 |   passwordRules: [String!]!
      3 | }

#### ❌ Error

    > 1 | type PasswordTooWeakError {
        | ^^^^^^^^^^^^^^^^^^^^^^^^^^^
    > 2 |   passwordRules: [String!]!
        | ^^^^^^^^^^^^^^^^^^^^^^^^^^^
    > 3 | }
        | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Types that end with 'Error' must implement UserError"
`;

exports[`Invalid #4 1`] = `
"#### ⌨️ Code

      1 | type PasswordTooWeakError implements OtherType {
      2 |   passwordRules: [String!]!
      3 | }

#### ❌ Error

    > 1 | type PasswordTooWeakError implements OtherType {
        | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    > 2 |   passwordRules: [String!]!
        | ^^^^^^^^^^^^^^^^^^^^^^^^^^^
    > 3 | }
        | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Types that end with 'Error' must implement UserError"
`;

exports[`Invalid #5 1`] = `
"#### ⌨️ Code

      1 | extend type NotErrorType implements UserError {
      2 |   passwordRules: [String!]!
      3 | }

#### ❌ Error

    > 1 | extend type NotErrorType implements UserError {
        | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    > 2 |   passwordRules: [String!]!
        | ^^^^^^^^^^^^^^^^^^^^^^^^^^^
    > 3 | }
        | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Types that do not end in 'Error' cannot implement UserError"
`;
